stardis/plasma/base.py

```diff
@@ -275,6 +273,139 @@ def calculate(
+            ]
+        ]
+
+        # Truncate to final atomic number
+        linelist = linelist[
+            linelist.atomic_number <= (atomic_data.selected_atomic_numbers.max())
+        ]
+
+        # Calculate energy of upper level
+        linelist["e_up"] = (
+            (
+                linelist.e_low.values * u.eV
+                + (const.h * const.c / (linelist.wavelength.values * u.AA))
+            )
+            .to(u.eV)
+            .value
+        )
+
+        exponent_by_point = np.exp(
+            np.outer(
+                -linelist.e_low.values * u.eV, 1 / (t_electrons * u.K * const.k_B)
+            ).to(1)
+        )
+
+        linelist_with_densities = linelist.merge(
+            ion_number_density / partition_function,
+            how="left",
+            on=["atomic_number", "ion_number"],
+        )
+
+        prefactor = (
+            exponent_by_point * linelist_with_densities[np.arange(points)]
+        ).values.T  # arange mask of the dataframe returns the set of densities of the appropriate ion for the line at each point
+
+        line_nus = (linelist.wavelength.values * u.AA).to(
+            u.Hz, equivalencies=u.spectral()
+        )
+
+        emission_correction = 1 - np.exp(
+            (
+                -const.h
+                / const.k_B
+                * np.outer(
+                    line_nus,
+                    1 / (t_electrons * u.K),
+                )
+            ).to(1)
+        )
+
+        alphas = pd.DataFrame(
+            (
+                ALPHA_COEFFICIENT
+                * prefactor
+                * 10**linelist.log_gf.values
+                * emission_correction.T
+            ).T
+        )
+
+        if np.any(np.isnan(alphas)) or np.any(np.isinf(np.abs(alphas))):
+            raise ValueError(
+                "Some alpha_line from vald are nan, inf, -inf " " Something went wrong!"
+            )
+
+        alphas["nu"] = line_nus.value
+        linelist["nu"] = line_nus.value
+
+        # Linelist preparation below is taken from opacities_solvers/base/calc_alpha_line_at_nu
+        # Necessary for correct handling of ion numbers using charge instead of astronomy convention (i.e., 0 is neutral, 1 is singly ionized, etc.)
+        ionization_energies = ionization_data.reset_index()
+        ionization_energies["ion_number"] -= 1
+        linelist = pd.merge(
+            linelist,
+            ionization_energies,
+            how="left",
+            on=["atomic_number", "ion_number"],
+        )
+
+        linelist["level_energy_lower"] = ((linelist["e_low"].values * u.eV).cgs).value
+        linelist["level_energy_upper"] = ((linelist["e_up"].values * u.eV).cgs).value
+
+        # Radiation broadening parameter is approximated as the einstein A coefficient. Vald parameters are in log scale.
+        linelist["A_ul"] = 10 ** (
+            linelist["rad"]
+        )  # see 1995A&AS..112..525P for appropriate units - may be off by a factor of 4pi
```

Comment:
I've checked the units on this again and added the factor in because I've convinced myself that's how the units work out. I've tried to check this against physical sources in a couple ways, but radiation broadening seems generally much smaller than van der waals or stark broadening so it's hard to see a difference in spectra, and comparing the coefficients themselves to those given by other databases doesn't work because they just have different values. I think dividing by 4 pi makes it a little better/closer between different sources, but I have no real justification for whether they should agree in the first place.